File: packages/core/src/automations/actions/run-command.ts
// packages/core/src/automations/actions/run-command.ts
//
// A1 (contract §6): script chips never touch shell text. The future
// run_action executor (Task 23) resolves each chip in the script to its
// plain string value but — unlike every other action's ChipText, which gets
// joined into one rendered string — keeps chip boundaries, passing them here
// as `script` parts. Each chip becomes its own MF_<n> child env var; only
// author-typed literal text ever becomes shell source.
import { execFile } from 'node:child_process';
import { existsSync } from 'node:fs';
import { z } from 'zod';
import { resolveAndValidatePath } from '../../server/routes/path-utils.js';
import type { ActionCtx, ActionDef } from './types.js';

const MAX_OUTPUT_BYTES = 8 * 1024 * 1024;
const STDERR_TAIL_CHARS = 4000;

const scriptPartSchema = z.union([z.object({ literal: z.string() }).strict(), z.object({ chip: z.string() }).strict()]);

export const RunCommandInputSchema = z
  .object({
    script: z.array(scriptPartSchema).min(1),
    runIn: z.enum(['project root', 'worktree', 'custom']),
    customPath: z.string().optional(),
    outputAs: z.enum(['text', 'lines']).optional(),
  })
  .strict()
  .superRefine((input, ctx) => {
    if (input.runIn === 'custom' && !input.customPath) {
      ctx.addIssue({ code: 'custom', message: "runIn 'custom' requires customPath" });
    }
  });

export type RunCommandInput = z.infer<typeof RunCommandInputSchema>;

/** Each chip becomes its own quoted `"$MF_<n>"` placeholder — the value never becomes shell text. */
export function compileScript(parts: RunCommandInput['script']): { script: string; env: Record<string, string> } {
  const env: Record<string, string> = {};
  let script = '';
  let chipIndex = 0;
  for (const part of parts) {
    if ('literal' in part) {
      script += part.literal;
      continue;
    }
    const name = `MF_${chipIndex}`;
    env[name] = part.chip;
    script += `"$${name}"`;
    chipIndex += 1;
  }
  return { script, env };
}

/** cwd is never shell source — 'custom' is the only mode that runs user-authored text through path containment (A1). */
function resolveCwd(ctx: ActionCtx, input: RunCommandInput): string {
  if (input.runIn === 'project root') return ctx.projectRoot;
  if (input.runIn === 'worktree') {
    if (!ctx.worktreePath)
      throw new Error('run_command runIn "worktree" requested but no worktree is active for this run');
    return ctx.worktreePath;
  }
  const resolved = resolveAndValidatePath(ctx.projectRoot, input.customPath!);
  if (!resolved) throw new Error(`run_command custom cwd '${input.customPath}' is outside the project root`);
  return resolved;
}

export function resolveShellBinary(exists: (path: string) => boolean = existsSync): string {
  return exists('/bin/zsh') ? '/bin/zsh' : '/bin/sh';
}

function extractExitCode(err: { code?: number | string | null } | null): number | null {
  if (!err) return 0;
  return typeof err.code === 'number' ? err.code : null;
}

function spawnScript(
  shell: string,
  script: string,
  cwd: string,
  env: Record<string, string>,
  signal: AbortSignal,
): Promise<{ exitCode: number; stdout: string; stderr: string }> {
  return new Promise((resolvePromise, rejectPromise) => {
    execFile(
      shell,
      ['-lc', script],
      { cwd, env: { ...process.env, ...env }, signal, maxBuffer: MAX_OUTPUT_BYTES },
      (err, stdout, stderr) => {
        const exitCode = extractExitCode(err);
        // A null exit code means the process never produced one (spawn error, abort) — that's not a script failure to report as exitCode.
        if (err && exitCode === null) {
          rejectPromise(err);
          return;
        }
        resolvePromise({ exitCode: exitCode ?? 0, stdout: String(stdout), stderr: String(stderr) });
      },
    );
  });
}

function formatOutput(stdout: string, outputAs: RunCommandInput['outputAs']): string | string[] {
  const trimmed = stdout.trim();
  if (outputAs !== 'lines') return trimmed;
  if (trimmed.length === 0) return [];
  return trimmed
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.length > 0);
}

export const runCommandAction: ActionDef = {
  id: 'run_command',
  title: 'Run command',
  group: 'builtin',
  auth: 'none',
  input: RunCommandInputSchema,
  outputs: [
    { name: 'output', type: 'text' },
    { name: 'exitCode', type: 'number' },
  ],
  idempotent: false,
  async run(ctx, rawInput) {
    const input = RunCommandInputSchema.parse(rawInput);
    const cwd = resolveCwd(ctx, input);
    const { script, env } = compileScript(input.script);
    const shell = resolveShellBinary();
    const { exitCode, stdout, stderr } = await spawnScript(shell, script, cwd, env, ctx.signal);
    if (exitCode !== 0) {
      const tail = stderr.trim().slice(-STDERR_TAIL_CHARS);
      throw new Error(`run_command exited ${exitCode}: ${tail || '(no stderr output)'}`);
    }
    return { output: formatOutput(stdout, input.outputAs), exitCode };
  },
};

File: packages/core-rs/crates/mainframe-automations/src/actions/run_command.rs
//! `run_command` (T6.3, Node actions/run-command.ts). A1 (contract §6):
//! script chips never touch shell text — each chip becomes its own `MF_<n>`
//! child env var and the script gets a quoted `"$MF_<n>"` where the chip
//! sat; only author-typed literal text is shell source. `cwd` is never
//! shell source either: `custom` is the one mode that runs user text, and
//! it goes through realpath containment against the project root.

use serde::Deserialize;
use serde_json::{Value, json};

use crate::domain::OutputAs;
use crate::tokens::TokenValue;

use super::manifest::{ActionAuth, ActionGroup, ActionManifest, ActionOutput, ActionOutputType};
use super::paths::resolve_and_validate_path;
use super::shell::{resolve_shell, spawn_script, tail_chars};
use super::{Action, ActionCtx, ActionError, ActionOutputs, parse_input};

const STDERR_TAIL_CHARS: usize = 4000;

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct LiteralPart {
    literal: String,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct ChipValue {
    chip: String,
}

/// `{literal: string} | {chip: string}` — the run_action verb keeps chip
/// boundaries for this one action instead of joining the ChipText (A1).
#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub(crate) enum ScriptPart {
    Literal(LiteralPart),
    Chip(ChipValue),
}

#[derive(Debug, Clone, Copy, Deserialize)]
pub(crate) enum RunIn {
    #[serde(rename = "project root")]
    ProjectRoot,
    #[serde(rename = "worktree")]
    Worktree,
    #[serde(rename = "custom")]
    Custom,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct RunCommandInput {
    script: Vec<ScriptPart>,
    run_in: RunIn,
    #[serde(default)]
    custom_path: Option<String>,
    #[serde(default)]
    output_as: Option<OutputAs>,
}

pub struct RunCommandAction;

impl Action for RunCommandAction {
    fn manifest(&self) -> ActionManifest {
        ActionManifest {
            id: "run_command",
            title: "Run command",
            group: ActionGroup::Builtin,
            auth: ActionAuth::None,
            credential_label_hint: None,
            params_schema: params_schema(),
            outputs: vec![
                ActionOutput::new("output", ActionOutputType::Text),
                ActionOutput::new("exitCode", ActionOutputType::Number),
            ],
            idempotent: false,
        }
    }

    fn execute<'a>(
        &'a self,
        params: &'a Value,
        ctx: &'a ActionCtx,
    ) -> crate::engine::BoxFuture<'a, Result<ActionOutputs, ActionError>> {
        Box::pin(async move {
            let input: RunCommandInput = parse_input("run_command", params)?;
            if input.script.is_empty() {
                return Err(ActionError(
                    "invalid input for 'run_command': script must not be empty".to_string(),
                ));
            }
            let cwd = resolve_cwd(ctx, input.run_in, input.custom_path.as_deref()).await?;
            let (script, env) = compile_script(&input.script);
            let shell = resolve_shell().await;
            let (exit_code, stdout, stderr) = spawn_script(&shell, &script, &cwd, &env).await?;
            if exit_code != 0 {
                let tail = tail_chars(stderr.trim(), STDERR_TAIL_CHARS);
                let detail = if tail.is_empty() {
                    "(no stderr output)"
                } else {
                    tail
                };
                return Err(ActionError(format!(
                    "run_command exited {exit_code}: {detail}"
                )));
            }
            let mut outputs = ActionOutputs::new();
            outputs.insert(
                "output".to_string(),
                format_output(&stdout, input.output_as),
            );
            outputs.insert("exitCode".to_string(), TokenValue::Number(exit_code as f64));
            Ok(outputs)
        })
    }
}

/// Each chip becomes its own quoted `"$MF_<n>"` placeholder — the value
/// never becomes shell text.
pub(crate) fn compile_script(parts: &[ScriptPart]) -> (String, Vec<(String, String)>) {
    let mut script = String::new();
    let mut env = Vec::new();
    for part in parts {
        match part {
            ScriptPart::Literal(p) => script.push_str(&p.literal),
            ScriptPart::Chip(p) => {
                let name = format!("MF_{}", env.len());
                script.push_str(&format!("\"${name}\""));
                env.push((name, p.chip.clone()));
            }
        }
    }
    (script, env)
}

/// cwd is never shell source — `custom` is the only mode that runs
/// user-authored text, through realpath containment (A1).
async fn resolve_cwd(
    ctx: &ActionCtx,
    run_in: RunIn,
    custom_path: Option<&str>,
) -> Result<String, ActionError> {
    match run_in {
        RunIn::ProjectRoot => Ok(ctx.project_root.clone()),
        RunIn::Worktree => ctx.worktree_path.clone().ok_or_else(|| {
            ActionError(
                "run_command runIn \"worktree\" requested but no worktree is active for this run"
                    .to_string(),
            )
        }),
        RunIn::Custom => {
            let requested = custom_path.ok_or_else(|| {
                ActionError("run_command runIn 'custom' requires customPath".to_string())
            })?;
            resolve_and_validate_path(&ctx.project_root, requested)
                .await
                .ok_or_else(|| {
                    ActionError(format!(
                        "run_command custom cwd '{requested}' is outside the project root"
                    ))
                })
        }
    }
}

fn format_output(stdout: &str, output_as: Option<OutputAs>) -> TokenValue {
    let trimmed = stdout.trim();
    match output_as {
        Some(OutputAs::Lines) => TokenValue::List(
            trimmed
                .split('\n')
                .map(str::trim)
                .filter(|line| !line.is_empty())
                .map(|line| TokenValue::Text(line.to_string()))
                .collect(),
        ),
        _ => TokenValue::Text(trimmed.to_string()),
    }
}

fn params_schema() -> Value {
    json!({
        "type": "object",
        "properties": {
            "script": {
                "type": "array",
                "minItems": 1,
                "items": {
                    "anyOf": [
                        {
                            "type": "object",
                            "properties": {"literal": {"type": "string"}},
                            "required": ["literal"],
                            "additionalProperties": false
                        },
                        {
                            "type": "object",
                            "properties": {"chip": {"type": "string"}},
                            "required": ["chip"],
                            "additionalProperties": false
                        }
                    ]
                }
            },
            "runIn": {"type": "string", "enum": ["project root", "worktree", "custom"]},
            "customPath": {"type": "string"},
            "outputAs": {"type": "string", "enum": ["text", "lines"]}
        },
        "required": ["script", "runIn"],
        "additionalProperties": false
    })
}

#[cfg(test)]
pub(crate) async fn resolve_cwd_for_test(
    ctx: &ActionCtx,
    run_in: &str,
    custom_path: Option<&str>,
) -> Result<String, ActionError> {
    let run_in: RunIn = serde_json::from_value(Value::String(run_in.to_string()))
        .map_err(|err| ActionError(err.to_string()))?;
    resolve_cwd(ctx, run_in, custom_path).await
}

// PORT STATUS: greenfield (docs/plans/2026-07-12-automations-v2-rust-engine.md T6.3), not a TS port
// confidence: high
// todos: 0
// notes: mirrors Node actions/run-command.ts (compileScript/resolveCwd/
//        spawnScript/formatOutput); 8 MB cap = Node's execFile maxBuffer,
//        enforced by capped stream reads + kill instead of maxBuffer.
